File: day-3/Part1.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Part1 {
    public static void main(String[] args) throws IOException {
        char[][] numbers = Files
                .lines(Paths.get(args[0]))
                .map(l -> l.toCharArray())
                .toArray(char[][]::new);
        int len = numbers[0].length;

        String gamma = IntStream
                .range(0, len)
                .mapToObj(digit -> Arrays.stream(numbers).filter(n -> n[digit] == '0').count() > numbers.length / 2
                        ? "0"
                        : "1")
                .collect(Collectors.joining(""));

        String epsilon = gamma
                .chars()
                .mapToObj(c -> c == '0' ? "1" : "0")
                .collect(Collectors.joining());

        System.out.println(Integer.parseInt(gamma, 2) * Integer.parseInt(epsilon, 2));
    }
}

File: day-3/part-1.fs
module Part1

[<EntryPoint>]
let main argv =
  let numbers = System.IO.File.ReadLines(argv.[0]) |> Array.ofSeq

  let gamma = (
    (seq {0..(numbers.[0].Length - 1)})
      |> Seq.map (fun i ->
        let count = numbers |> Array.filter (fun n -> n.[i] = '0') |> Array.length
        if count > numbers.Length / 2 then '0' else '1'
      )
      |> Array.ofSeq
      |> System.String
  )

  let epsilon = (
    gamma
      |> Seq.map (fun n -> if n = '0' then '1' else '0')
      |> Array.ofSeq
      |> System.String
  )

  ((System.Convert.ToInt32 (gamma, 2)) * (System.Convert.ToInt32 (epsilon, 2)))
    |> printfn "%A"

  0
